feat(new source): Add host metrics source

bruceg opened the pull request:
This implements the host metrics source, as specified in https://github.com/timberio/vector/blob/master/rfcs/2020-08-26-3191-host-metrics.md

It is missing the following metrics due to missing support in the heim crate:

- `host_disk_read_errors_total` tagged with the disk (counter)
- `host_disk_read_retries_total` tagged with the disk (counter)
- `host_disk_read_sectors_total` tagged with the disk (counter)
- `host_disk_read_time_seconds_total` tagged with the disk (counter)
- `host_disk_write_errors_total` tagged with the disk (counter)
- `host_disk_write_retries_total` tagged with the disk (counter)
- `host_disk_write_time_seconds_total` tagged with the disk (counter)
- `host_disk_written_sectors_total` tagged with the disk (counter)
- `host_filesystem_avail_bytes` tagged with the device, filesystem type, and mountpoint (gauge)
- `host_filesystem_device_error` tagged with the device, filesystem type, and mountpoint (gauge)
- `host_filesystem_size_bytes` tagged with the device, filesystem type, and mountpoint (gauge)
- `host_filesystem_total_file_nodes` tagged with the device, filesystem type, and mountpoint (gauge)
- `host_filesystem_free_file_nodes` tagged with the device, filesystem type, and mountpoint (gauge)
- `host_network_receive_multicast_total` tagged with device (counter)
- `host_network_transmit_multicast_total` tagged with device (counter)

Commit:
feat(new source): Add host metrics source

Currently only outputs CPU times.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add basic memory info metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add basic swap info

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Use a macro to reduce boilerplate

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add load average gauges

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add non-Windows swapped memory info

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add MacOS extended swap info

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Tag metrics with their collector and host

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move run function into struct HostMetricsConfig

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for filtering on specified collectors

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add counters for network metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for selecting network metrics by device name

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add filter_result helper function

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add basic filesystem metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for pattern matches on network and filesystem filters

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for disk metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add support for a namespace option

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fill in linux memory metrics extensions

Signed-off-by: Bruce Guenter <bruce@timber.io>

jamtur01 commented:
Awesome! Does the heim crate have a bug or just doesn't support those - is there any plan to support them?

Commit:
Add support for negating patterns

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
> Awesome! Does the heim crate have a bug or just doesn't support those - is there any plan to support them?

No, the crate just doesn't support them. I don't see a plan to add support, but I would think PRs would be welcome.

Commit:
Fix clippy lints

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add tests and fix some bugs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move a test helper function out of the group of tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Tiny tweak: Use .join instead of format! for speed

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix a clippy lint

Signed-off-by: Bruce Guenter <bruce@timber.io>

jszwedko commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+    }
+}
+
+fn vec_contains_str(vec: &Option<Vec<PatternWrapper>>, value: &str) -> Option<()> {
```
I see this essentially `or`s the filters together?

I'm wondering if we wouldn't expect negations to filter out values that previously matched though. For example, if I have a filter `[["sd*", "!sda"]` would the user expect "sda" to be rejected? I personally would have.

Or, if I had `["sd*", "!dm-*"]` I might have expected "xvda" to be rejected, but I think this would allow that through as it matches "!dm-*".

We may want to structure it as `or`ing all positive filters (i.e. not negations) and then applying the negations; however, that would require a filter set like `["*", "!dm-*"]` to literally grab anything that isn't `dm-*`.

We could also consider providing a simple filter syntax that would let people write things like `"sd*" && !sda && !dm-*" . I'm reminded of `tcpdump`'s filter syntax (e.g. `tcpdump -i eth1 port not 22 and port 53 and host 1.2.3.4`). That would require a bit more heavy lifting though :smile: . We could also just drop negations for now and only allow positive matches.

bruceg replied:
This does "or" all the filters together. The first one to match passes the filter. I'm not excited about making the filter logic more complex without a demonstrated use case, as explaining how it works will be a pain.

One alternative would be to split the filter lists into pairs of lists. That is:
```toml
disk.devices.include = ["d*"]
disk.devices.exclude = ["dm-*"]
```

That would be easy to explain and straightforward to implement. I'm not particularly proud of the filter negation mechanism, but it was part of the RFC. @jamtur01 any opinion?

jamtur01 replied:
The alternative seems reasonable. I guess it's easier to document?

jszwedko replied:
:+1: I like splitting it up into `include` and `exclude`. That seems relatively simple from  the user's perspective.

jszwedko reviewed (commented):
This is looking good! I like the pattern wrapper stuff.

My main concern is around the negation matching for filters (see comment).

I attached the output I saw when running it just for reference: [metrics.txt](https://github.com/timberio/vector/files/5308476/metrics.txt)

jszwedko commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+                    if let Some(device) = partition.device() {
+                        tags.insert("device".into(), device.to_string_lossy().into());
+                    }
+                    stream::iter(
```
Ah, never mind, I missed that this is just for one filesystem and then it is flattened later.

bruceg replied:
Do you think there any benefit in moving some of the internals out (ie the creation of the intermediate stream) into another function to reduce the complexity?

jszwedko replied:
It's up to you. I think it is fairly readable as-is. I appreciated the comments.

bruceg commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+    }
+}
+
+macro_rules! tags {
```
Ah, I had written the `btreemap` macro first, then got frustrated with all the `.to_string` and `.into` calls, and wrote `tags!`. Obviously I missed converting them all. I'll drop one.

bruceg commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+
+    fn has_collector(&self, collector: Collector) -> bool {
+        match &self.collectors {
+            None => true,
```
This pattern matches the behavior of the filter lists, though (without globs, of course). I'm inclined to keep it this way until we need non-default collectors.

bruceg commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+                        memory.available().get::<byte>() as f64,
+                        tags![],
+                    ),
+                    #[cfg(target_os = "linux")]
```
They are only exposed by the `heim` crate on Linux, so no. The ones marked as "not windows" are similarly exposed. The bounds all come from heim.

jszwedko replied:
Gotcha, I'm happy to match the `heim` constraints.

bruceg commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1100 @@
+                    .await
+            }
+            Err(error) => {
+                error!(message = "Failed to load network I/O counters", %error, rate_limit_secs = 60);
```
Possibly. I think they can only be caused by local configuration issues (ie permissions on `/proc` or the like) or bugs in the heim crate. That doesn't seem to be a strong indicator for adding events, but I'll be happy to write them up.

jszwedko replied:
:+1: 

Should we have any internal events in this source? I didn't see any. Maybe just one that is emitted during each scrape interval?

bruceg replied:
I can do that.

bruceg commented on .meta/sources/host_metrics.toml.erb:
```diff
@@ -0,0 +1,104 @@
+function_category = "collect"
+output_types = ["metric"]
+requirements = {}
+strategies = ["daemon", "sidecar", "service"]
```
I don't actually know. I searched for some time what this should be, but found no reference to guide me.

jszwedko replied:
https://vector.dev/docs/setup/deployment/strategies/ has some more details on them. For host metrics, I feel like `daemon` is the only relevant one.

bruceg replied:
It still seems like `sidecar` could be valid if vector is running on the same host, and `service` can give metrics on the host that is running vector.

jszwedko replied:
Hmm, I'm not quite sure what that would look like. Would the service be exposing `/proc` and the like to the `vector` container?

I tend to think of the sidecar as vector ingesting logs/metrics _from_ the service _about_ that service.

bruceg replied:
Ah, I see what you mean. Ok, I'll fix that.

Commit:
Drop btreemap macro, everything should have used tags! instead

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rework the filter lists into separate include/exclude lists

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Promote a bool::and_some feature to crate level

A similar feature was used in both the host metrics source and TLS
settings, so unify it centrally.

Signed-off-by: Bruce Guenter <bruce@timber.io>

jszwedko reviewed (approved):
Nice work!

jszwedko commented on .meta/sources/host_metrics.toml.erb:
```diff
@@ -0,0 +1,94 @@
+  term: "mountpoint path",
+  metrics: "utilization",
+  examples: [["eth0"], ["enp?s*"]]
+) %>
```
Actually, I realized it'd be nice to list out all of the metrics collected for each OS in the docs somewhere.

bruceg commented on .meta/_partials/fields/_includes_excludes.toml:
```diff
@@ -0,0 +1,22 @@
+required = false
+examples = <%= examples.to_toml %>
+description = """\
+The list of <%= term %> patterns for which to gather <%= metrics %> metrics. \
```
It is indeed a "glob" style filename match, and no it isn't documented anywhere. I'll add a note to this partial.

bruceg commented on .meta/_partials/fields/_includes_excludes.toml:
```diff
@@ -0,0 +1,22 @@
+examples = <%= examples.to_toml %>
+description = """\
+The list of <%= term %> patterns for which to gather <%= metrics %> metrics. \
+Defaults to including all <%= term %>s.\
```
I guess. My thinking was that making them options skips a couple steps of indirection in the common case (no iterator, no call to fnmatch). However, it does complicate the matching code, and handling them is not a hot path (unless the system has a huge number of devices, but even then it is polled at most once a second). I'll switch.

bruceg replied:
Ah, there's a gotcha here. For filesystem metrics, the filesystem may have no device. If no patterns are specified, this filesystem currently would be included, but any device pattern would allow the user to chose to exclude them. Forcing a pattern match would require us to either always include or always exclude such filesystems. Now, these filesystems are *probably* not interesting for metrics, but it is a consideration.

jszwedko replied:
:thinking: I think I'd expect "no device" to also match `*` though so, if I literally specified `[ "*" ]` as the config, I'd still get a match. I concede that is a bit of an odd case.

Commit:
Fix default unwrap for device-less filesystems

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix the installation strategies

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add an internal event counting the number of metrics generated

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add mention of glob pattern matching to docs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen commented on src/sources/host_metrics.rs:
```diff
@@ -0,0 +1,1189 @@
+            metrics.extend(add_collector("memory", self.memory_metrics().await));
+            metrics.extend(add_collector("memory", self.swap_metrics().await));
+        }
+        if self.has_collector(Collector::Network) {
+            metrics.extend(add_collector("network", self.network_metrics().await));
+        }
```
It's probably not worth changing now, but as we add more of these we could think about moving these collection methods onto the collectors themselves, so you could do something like

```rust
for collector in self.collectors {
    metrics.extend(collector.collect().await);
}
```

That'd also let us be a bit more modular as we implement more types of metrics.

bruceg replied:
How would that work? `collectors` is `Vec<enum Collector>`, can you define methods on individual variants of an enum?

lukesteensen replied:
It involve switching from an enum to a trait with `typetag` so that it can be deserialized. Similar to what we do with sources/sinks/transforms. Most of the enum solutions I can think of involve some degree of boilerplate.

lukesteensen reviewed (approved):
Looks good to me!

Commit:
Add CPU index to CPU metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Unify one metric duplicated between Linux and MacOS

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Document the list of output metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix documentation validation typos

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix a couple cross-OS portability issues

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented:
Looks great @bruceg, nice work!

Commit:
Don't run load average tests on Windows

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic commented:
@bruce as part of this can we add a `docs/reference/components/sources/host_metrics.cue` file instead? Thanks.

Commit:
Add mount point detail to partition usage load error

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Be more permissive in lengths of filtered results

The tests for filtering metrics assume there are more than one unique
value for the given tag. This is not necessarily true, particularly in
restricted environments, so relax the asserted inequalities.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Check that the filtered results add up

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Cancel two more tests on Windows to make this source pass CI

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Start transitioning docs to new metadata (incomplete)

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Run `cue fmt`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Finish metadata documentation

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add owner of docs

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
@binarylogic I think the cue file is done, please double check before I merge.

binarylogic reviewed (approved):
Looks great, nice work!

Commit:
Merge remote-tracking branch 'origin/master' into host-metrics-source

Signed-off-by: Bruce Guenter <bruce@timber.io>
